internal/command/run/run.go

```diff
@@ -21,55 +22,76 @@ import (
 	"github.com/weaveworks/flintlock/internal/version"
 	"github.com/weaveworks/flintlock/pkg/defaults"
-	"github.com/weaveworks/flintlock/pkg/flags"
 	"github.com/weaveworks/flintlock/pkg/log"
 )
 
 // NewCommand creates a new cobra command for running flintlock.
-func NewCommand(cfg *config.Config) (*cobra.Command, error) {
-	cmd := &cobra.Command{
-		Use:   "run",
-		Short: "Start running the flintlock API",
-		PreRunE: func(c *cobra.Command, _ []string) error {
-			flags.BindCommandToViper(c)
-
-			logger := log.GetLogger(c.Context())
-			logger.Infof(
-				"flintlockd, version=%s, built_on=%s, commit=%s",
-				version.Version,
-				version.BuildDate,
-				version.CommitHash,
-			)
-
-			return nil
-		},
-		RunE: func(c *cobra.Command, _ []string) error {
-			return runServer(c.Context(), cfg)
+func NewCommand(cfg *config.Config) *cli.Command {
+	cmd := &cli.Command{
+		Name:  "run",
+		Usage: "Run flintlock",
+		Action: func(c *cli.Context) error {
+			return runServer(c.Context, cfg)
 		},
 	}
 
 	cmdflags.AddGRPCServerFlagsToCommand(cmd, cfg)
+	cmdflags.AddContainerDFlagsToCommand(cmd, cfg)
+	cmdflags.AddFirecrackerFlagsToCommand(cmd, cfg)
+	cmdflags.AddNetworkFlagsToCommand(cmd, cfg)
+	cmdflags.AddHiddenFlagsToCommand(cmd, cfg)
+
+	stateRootDirFlag := altsrc.NewPathFlag(&cli.PathFlag{
+		Name:        "state-dir",
+		Usage:       "Path to the directory where flintlock will store its state",
+		Value:       defaults.StateRootDir,
+		Destination: &cfg.StateRootDir,
+	})
+
+	resyncPeriodflag := altsrc.NewDurationFlag(&cli.DurationFlag{
+		Name:        "resync-period",
+		Usage:       "Resync period for the specs reconciliation",
+		Value:       defaults.ResyncPeriod,
+		Destination: &cfg.ResyncPeriod,
+	})
+
+	cmd.Flags = append(cmd.Flags, stateRootDirFlag, resyncPeriodflag)
+
+	cmd.Before = func(context *cli.Context) error {
+		// Load the configuration file
+		configPath := context.String("config")
+		if _, err := os.Stat(configPath); err == nil || context.IsSet("config") {
+			inputSource, err := altsrc.NewYamlSourceFromFile(configPath)
+			if err != nil {
+				return fmt.Errorf("unable to create input source with context: %w", err)
+			}
 
-	if err := cmdflags.AddContainerDFlagsToCommand(cmd, cfg); err != nil {
-		return nil, fmt.Errorf("adding containerd flags to run command: %w", err)
-	}
-
-	if err := cmdflags.AddFirecrackerFlagsToCommand(cmd, cfg); err != nil {
-		return nil, fmt.Errorf("adding firecracker flags to run command: %w", err)
-	}
-
-	if err := cmdflags.AddNetworkFlagsToCommand(cmd, cfg); err != nil {
-		return nil, fmt.Errorf("adding network flags to run command: %w", err)
-	}
+			err = altsrc.ApplyInputSourceValues(context, inputSource, cmd.Flags)
+			if err != nil {
+				return fmt.Errorf("unable to apply input source with context: %w", err)
+			}
+		}
 
-	if err := cmdflags.AddHiddenFlagsToCommand(cmd, cfg); err != nil {
-		return nil, fmt.Errorf("adding hidden flags to run command: %w", err)
+		if cfg.ParentIface == "" {
```

Comment:
Does `urfave/cli` handle required fields and raising an error if its missing?

Reply:
yes there is `required: true`, but the way it behave is that you have to set the flag manually, I think the check happens before the `command.before` where we set the flag with the config file, causing it to fail

Reply:
is it possible to set a config file and flags at the same time? which takes precedence?

Reply:
The precedence for flag value sources is as follows (highest to lowest):

1. Command line flag value from user
2. Environment variable (if specified)
3. Configuration file (if specified)
4. Default defined on the flag